Paper: NEXT

stsievert opened the pull request:
* Title: "NEXT: A system to easily connect crowdsourcing and adaptive data collection"
* Authors: Scott Sievert, Daniel Ross, Lalit Jain, Kevin Jamieson, Rob Nowak, Bob Mankoff
* Keywords: crowdsourcing, adaptive sampling, system

sbenthall commented:
I'm excited to say I'll be reviewing this paper.

AngelosPsy commented:
<b> Next: A system to easily connect crowdsourcing and adaptive data collection 

Review
</b>

I would like to thank the authors for all their impressive work. I hope that my comments below are helpful during the revision process of the paper. 

The authors do a good work in presenting their software using real life examples and applications (e.g., the work that NEXT does for the New Yorker data). As the manuscript, though, refers to a multidisciplinary audience, I found out that some of the used terms should have been better explained. I provide my specific comments below.

<b> Written quality </b>

<i> Title </i>
I do not think that the link to the Monty Python video is necessary.


<i> Abstract </i>

<ul>
 
<li>`We have created...algorithms`: I think it is better to first introduce the problem (here the problems in collecting crowdsourced data) and then say how you are going to solve it (by creating NEXT). </li>

 <li>`crowdsoured data` should be crowdsourced data</li>

 <li>'Adaptive sampling uses previous responses to collect
data and produces accurate results with minimal samples.' This sentence is really hard to understand. The authors could make it simpler by being a bit more specific. For example, when they mention 'accurate results' you could mention to what these results refer to (e.g., categorization of images to funny or unfunny as mentioned later in the main text).</li>

<li>'practical gains...system challenges'. I would give one example for the gains and one example for the challenges. </li>

<li>'We will explain...and API.' I think that this sentence would fit better at the end of an introduction section. At the end of the abstract I would probably try and reach a general conclusion  (e.g., our software is important because X and Y).</li>

<i>Introduction</i>

<li>'For example... internet bots' I think this sentence needs to be rewriten, as it is not crystal clear how the action of selecting certain objects protects from bots. For example, the authors could write that a service is protected by internet bots by using this image selection task since internet bots, in comparison to human agents, have difficulty in completing the task accurately. </li>

<li>The second paragraph needs to be rewritten as many sentences are not specific enough. To illustrate, it is written that 'Crowdsourcing enables the collection of a large amount of
data, which is necessary to extract complex results from simple human judgments.' It should be specified what type of data (e.g., decision making data) and what makes the said results complex. The authors provide an example after this sentence but I did not understand why the provided results (e.g., shoe similarity) are complex. In a similar vein, on the last line it is not clear what crowdsourcing exactly does.</li>
	
 <li>I reckon it would enhance the flow of the paper if at the end of the introduction, there was a short paragraph in which the skeleton of the rest of the paper was introduced.</li>

<i>Problem statement</i>

<li>'The standard machine learning workflow obtains human labels before producing a model that predicts labels.' I think this sentence could have been a bit clearer. For example, 'The standard machine learning workflow obtains labels of human data before producing a model that predicts labels of future data.' This is of course an example that they authors may use or come up with a better one.  </li>

<li>It would have been better if Figure 1 was placed on the same page as the text that it refers to. </li>

  <li>Please refer to some studies that simulated experiments (see p. 2, paragraph 2).</li>

<i>Our System</i>

  <li>I think it is enough if it is mentioned that NEXT was designed in UW-Madison only once in the whole text.</li>
  
   <li>'Various adaptive algorithms have been evaluated in the real world, verifying expected efficiency gains.' Please provide references or links to example results.</li>

 <i> Applications of NEXT</i>

  <li>After you introduce the applications, but before you go into one of them, maybe the authors could write something like: 'We now describe each application in detail'. </li>


<i> Algorithm implementation </i>

  <li>'The arguments and return values for all algorithm functions are specified exactly, in a YAML-based schema.' Remove comma.</li>

  <li>'...and contains four root level keys for each of initExp, getQuery, processAnswer and getModel.'Add comma before 'and getModel'.</li>


<b> Technical Content </b>
  <li>I think that it would be better if the foo example in the Algortithm implentation section, would have been replaced by a real example that refers back to the text. This goes back to the example used in the 'Database access' part. Specifically in this part, it is difficult to follow what the goal of the code is.</li>

<b> Novelty and impact </b>

  <li>The authors make a convincing case as to why adaptive algorithms should be used instead of passive ones.</li>

<b> Verifiability </b>

<li>This section relates to the 'Technical Content' section so I refer to it</li>

<b> Other requirements</b>

<b> Figures </b>
  <li>Do the authors have the rights to reproduce all included figures?</li>
  <li>Caption Figure 6. Delete second full stop.</li>
<li>A reference to PyMongo should be added.</li>
</ul>

stsievert commented:
@AngelosPsy thanks for the review! It's been helpful – it's pointed out issues.

I've rewritten the abstract, revised the introduction and provided more complete examples. Complete changes are in the full PDF.

> I do not think that the link to the Monty Python video is necessary.

It's included by default 😜. It'll link to the talk, after I give it.

> Do the authors have the rights to reproduce all included figures?

Thanks – I'll check.

FYI, GitHub allows you to make comments directly on the .rst file. After clicking on "File changed", a little "+" button appears when hovering over the document.

sbenthall commented:
**Introduction**

- The opening example is difficult to follow. You begin by bringing up crowdsourcing but then describe a CAPTCHA system, which is not the same thing as crowdsourcing. You might want to cite a source on the phrase 'wisdom of the crowd', and even use that source in your motivation.

- You second example, about the shoes, is also hard for me to follow. What is a "low working memory constraint"? A small fraction of the available shoes are used...for what?

- I see a reference to an earlier paper on NEXT. I'm wondering what differentiates this paper from it. Explaining that difference in focus could help the reader understand the contribution made by this paper.

**Problem statement**

- The kind of machine learning problem you describe in the first paragraph is supervised classification. You should use those words. There must be some [original sources](https://scholar.google.com/scholar?hl=en&q=supervised+classification) on this topic you could cite. 

- It now reads like adaptive data collection is what your paper is mainly about. If NEXT is a tool for adaptive data collection, it would be great to say this in your introduction.

- Is there a canonical citation for adaptive data collection? who had this idea first?

- There is an oddness to the way grammatical tense and the logic of your argument connect. You are saying adaptive data collection has improved through experimentation. But this experimentation depends on the ease with which machine learning researchers and experimentalists can both work on the problem. Has this condition already been met? I think with some adjustment this can be cleared up.

-  I'm not sure what the Introduction is doing at this point; I didn't really understand where the paper was going until this section. I think it might be best to begin with some of the motivation here: that you are interested specifically in the use of adaptive sampling/algorithms in applications having to do with crowdsourcing. Or something like that. Clarity around this will help me understand what this software is good for.

- In the 'Challenges' section, I don't understand what you are getting at exactly. What is the difference between adaptive data collection and adaptive algorithms? Are they the same thing? How are 'the most useful queries' selected, given previous inputs? Is it to maximize information gain or something? Can you provide any more citations to work showing challenges raised by adaptive data collection in mathematics, systems, etc? How are the challenges of designing such a system different from the design of any interactive web application that responds to user data? What are the goals of NEXT that distinguish it from Microsoft's Decision Service.

I'm excited to read and review the rest of the paper soon. It looks like very interesting and important work. For the sake of the quality of the scholarship, I recommend that the writing and use of citations in these first sections can be improved.

sbenthall commented:
This is a cool paper. Here is more feedback.

**Our System**

- "In our usage, the system remains responsive to participants even after ...." It would be helpful to know more specifics of the system in order to understand how to interpret this claim. In particular, I'm not sure what parts of the system NEXT is responsible for, and what parts might be configurable and delegated to an IT backbone. Is this on the cloud? Is it backed by an SQL database? Or is this a comment on the run-time of the adaptive algorithms?

**Applications of NEXT**

- What is "adaptive gain"? How is it measured?

sbenthall commented:
Regarding your **Fig. 2**:

I am not sure what is being communicated by this figure. Maybe if you had a second figure illustrative a non-adaptive algorithm that you could contrast with it, it would be clearer why this is a new kind of approach.

The other question this figure raises for me is where data storage occurs. Is the algorithm making adaptive decisions based only on the answers of a single participant, or are the algorithmic decisions being made in the context the collected data from many users? If the latter, then is the algorithm using this collected data in real-time?

sbenthall commented:
**NEXT Architecture**

"That means these algorithms mirror the implementation in academic papers" -- I don't know what this means in this context. Which papers? What is that implementation?

Why are arguments and return values being specified in YAML? Why wouldn't you write the algorithm in Python or some other programming language?

Or is this just the way you are specifying your API in a machine-readable way? Is it documentation? What happens when you change the schema? What flexibility does this way of recording function schemas offer?

Database access: What kinds of databases are compatible with NEXT? The two examples are NoSQL databases. Is this important? Why?

Typo: "evaulation [sic] of different algorithms"

What functionality does the code example (including the class MyAlg) illustrate? I don't understand why I would want to run this code.

sbenthall commented:
**Experimentalist use***

Suppose I follow these instructions. How can I, as an experimentalist, access the results of the experiment for downstream data processing. I see that the results are displayed in the dashboard. Can they be downloaded from there?

I'd be interested to see more of a demonstration of what you can do with the results of such an experiment. I guess the use case is selecting the funniest caption. Is there a way to understand the distribution of funniness one gets as a result of the data? Or is this left to some other analyst to provide?

Given the availability of the experiment at an endpoint, how do the authors recommend distributing the URL to experimental subjects? Do the adaptive algorithms used compensate at all for any limitations or irregularities that may arise from sampling?

**Conclusion**

You claim that the system has been widely used by experimentalists in a lot of different settings. Maybe I missed the evidence for this in my reading, but I see the New Yorker cartoon case but not much else documented. Are there any citations available for other work that uses NEXT?

Thanks for writing this paper! It's an exciting piece of technology to know about and I will definitely keep it in mind if I can think of any applications that may use it. I hope these comments are valuable in identifying where this reviewer found room for improvement in the presentation of the technology.

stsievert commented:
@sbenthall thanks for the review! Your comments have been helpful, and I've taken a first pass at addressing them. Here's replies to a selection of comments:

>  Are there any citations available for other work that uses NEXT?

I included a paragraph about this in the section "Our system". I've expanded this paragraph, and am leaving the sentence in the conclusion.

>  Is this on the cloud? Is it backed by an SQL database? 

Thanks for this, it's needed. I've added a section on what NEXT actually is in "Our system" below the bulleted list.

> Regarding your Fig. 2:

I've expanded this figure to include a comparison with passive algorithms (and storage too).

AngelosPsy commented on papers/scott_sievert/next.rst:
```diff
@@ -111,15 +112,15 @@ access to unprecedented amounts of human judgment data.  Tapping into the
 an image that contain a certain object (e.g., "select all image locations that
-contain buildings") to protect against internet bots.
+contain buildings") to protect against internet bots. This task is trivial for
+a human but difficult a computer (or internet bots).
```
I think it should be "...but difficult _for_ a computer..."

AngelosPsy commented:
@stsievert Thank you for addressing most of my comments. Could you please let me know when you have addressed the last ones (e.g., the copyright of the figures).

Thank you again!

sbenthall commented:
@mpacer Just curious what the "needs more review" tag means. Is there anything else this paper needs before the June 21st deadline? Or are we good to go until final reviews?

stsievert commented:
@AngelosPsy One of the co-authors (Lalit Jain) met with The New Yorker (TNY). After their meeting in New York during NIPS 2015, he sent an email:

> **To:** Robert Nowak, Kevin Jamieson, Daniel Ross, Scott Sievert
> **From:** Lalit Jain
> **Date:** December 8, 2015 at 4:22 PM
> Hi All
> 
> The New Yorker meeting went well. They are excited to use NEXT for the caption contest.
> 
> They also gave us permission to release the data. Can you go ahead and do that and then link to it on the main webpage? The one caveat was giving the artist and the New yorker full credit. Please do this in a text file?
> 
> Lalit

I've given TNY and the individual artists credit in my paper, presentation and in the data repo.

AngelosPsy commented:
@stsievert Thank you for that. I am happy with the changes and thank you very much for your hard work. I am satisfied with your answers to my comments so for, for my part, the paper is good to go.

@proceedings, The paper is ready to be merged.

stsievert commented:
Thanks @AngelosPsy! I am sending the paper around another round of comments/edits. @sbenthall, do you have any more comments?

stsievert commented:
@AngelosPsy @sbenthall I've made some more edits, and think the paper is "camera-ready".

sbenthall commented:
The revisions are great. The paper is much improved.
Thanks again for the hard work to write this paper and bring this cool software to the SciPy community.

I think the paper could be merged.

I do have a few very minor comments as suggestions for improvement.
I'm aware it's very close to the final deadline for decisions and I definitely don't want to block merging this paper!

But I do think if it's possible to address these minor points, it would improve the paper.

> "Supervised  machine  learning  relies  on  human  responses  to  help
> produce  a  model  that  can  predict  the  response  a  human  would give." 

It looks like you are providing a definition of supervised machine learning here.
But what you are actually doing is discussing a particular application of supervised machine learning.
(For example, you could used supervised machine learning in case where labels were not provided by a humans, but were provided as part of some other more automated data flow pipeline.)
Could you make that clearer?

> Adaptive data collection is a process which selects
> the most useful data (in some sense) as quickly as possible to help
> achieve some goal (e.g., classification accuracy)

"(in some sense)" here is throwing me off I think.
Is there a specific sense you mean?
Isn't the sense defined in the rest of the sentence?

> The  simple  judgments humans  can  provide  require  many  responses.  A  simple  example
> is with using comparisons of two items to sort a list of n items.

I really like this example but I find your phrasing here a little awkward. What about:
"An example of a simple task requiring many human judgments is sorting a list of n items based on pairwise judgments of human preference for the items."
Or something like that.

**Fig 2. and Fig 3**

I think that the box labeled 'Results' in Fig 2 and the box labeled Results in Fig 3 must mean different things; otherwise the diagrams are inconsistent. Consider changing the label of the Results box in Fig 2 to "Model"?

stsievert commented:
Thanks for the feedback @sbenthall! I've addressed all your comments.

> equiring many human judgments is sorting a list of n items

I've rewritten that paragraph, let me know what you think. I don't want to make any claims about gains seen in crowdsourcing with sorting.

> Fig 2. and Fig 3 ... otherwise the diagrams are inconsistent.

Thanks for that – I didn't notice! I've removed the Results block from Fig. 3 and changed the caption to 

> The system required to use adaptive algorithm with crowdsourcing. The results are stored in the model, which may contain additional information.

mpacer commented:
Merging this for the purposes of having the camera-ready proceedings available for the main conference. Further changes (if necessary) should be made as a new PR.

stsievert commented:
I'd like to link to this paper in my presentation (I'll mention it several times). Is there a PDF available, should I distribute a draft, or should I point this repo?

mpacer commented:
There should be an official link soon. It won't be the final version(no
video links) but it should otherwise be correct.

On Tue, Jul 11, 2017 at 11:53 Scott Sievert <notifications@github.com>
wrote:

> I'd like to link to this paper in my presentation (I'll mention it several
> times). Is there a PDF available, should I distribute a draft, or should I
> point this repo?
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/272#issuecomment-314506157>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACXg6GgjOwdcTYqJtYBLCovlbm4j-DfZks5sM6iUgaJpZM4No0GE>
> .
>
